README.md
[<img width="200" alt="get in touch with Consensys Diligence" src="https://user-images.githubusercontent.com/2865694/56826101-91dcf380-685b-11e9-937c-af49c2510aa0.png">](https://diligence.consensys.net)<br/>
<sup>
[[  🌐  ](https://diligence.consensys.net)  [  📩  ](mailto:diligence@consensys.net)  [  🔥  ](https://consensys.github.io/diligence/)]
</sup><br/><br/>


# Inline Bookmarks

<img width="350" alt="inline_bookmarks_icon" src="https://user-images.githubusercontent.com/2865694/69679883-91366500-10a9-11ea-8e36-ba3e799127b3.png">

Code navigation for inline bookmarks made easy.

* Specify words inside your documents that are highlighted as bookmarks in the IDE. 
* Easily navigate to bookmarks from the **Inline Bookmarks View** added to the Explorer Panel. 

Other extensions allow you to set bookmarks with some clicks in the IDE. With Inline Bookmarks you specify custom trigger words inside your documents that are automatically turned into VSCode Bookmarks. They will show up with a customizable gutter icon next to the code lines and are highlighted within the code. Decorations can be customized. An **Inline Bookmarks** View in the Explorer Panel helps you keep track of all your bookmarks. Navigating to the bookmark location is as easy as clicking on an item in the view.

The extension can be used to keep track of development notes and todo's, for document review, as well as log analysis and auditing purposes. Since bookmarks are stored with your documents you can easily share your notes and bookmarks with others.

## Tour

<img width="350" alt="inline_bookmarks_icon" src="https://user-images.githubusercontent.com/2865694/69009240-d5ac5d00-0953-11ea-811e-f7fded9c6ecb.png">

![vscode-inline-bookmarks-2](https://user-images.githubusercontent.com/2865694/69681775-67803c80-10af-11ea-8e99-c79caf7781a5.gif)


#### Customizable Inline Bookmarks

* the following bookmark styles are provided by default: red, green, blue, purple

    <img width="350" alt="inline_bookmarks_icon" src="https://user-images.githubusercontent.com/2865694/69679883-91366500-10a9-11ea-8e36-ba3e799127b3.png">

**Note:** Additional bookmark types with custom colors/decoration styles can be added in the extension settings. See [FAQ](#FAQ).

* the following default trigger words/tags are configured by default:
  * `@todo` - (blue) General ToDo remark.
  * `@note` - (blue) General remark.
  * `@remind` - (blue) General remark.
  * `@follow-up` - (blue) General remark.
  * `@audit` - (red) General bookmark for potential issues.
  * `@audit-info` - (blue) General bookmark for information to be noted for later use.
  * `@audit-ok` - (green) Add a note that a specific line is not an issue even though it might look like.
  * `@audit-issue` - (purple) Reference a code location an issue was filed for.

**Note:** Words can be fully customized in: `code → preferences → Extensions: Inline Bookmarks`.
**Note:** Keyboard shortcuts can be assigned in: `code → preferences → keyboard shortcuts`.

<img width="350" alt="inline_bookmarks_icon" src="https://user-images.githubusercontent.com/2865694/69680065-1457bb00-10aa-11ea-853b-c36fdfc68b96.png">

#### Filterable Bookmarks Finder

command: `inlineBookmarks.showSelectBookmark`

<img width="350" alt="inline_bookmarks_icon" src="https://user-images.githubusercontent.com/2865694/81169692-8b6a7700-8f99-11ea-9ab2-8b121154c99a.gif">

#### Bookmarks View

* click on a bookmark to jump to its location
* click refresh to re-scan loaded files for changes
* toggle to only show bookmarks for visible editors
* the bookmarks view will follow your cursor location in the editor. the follow mode can be changed with the configuration option `inline-bookmarks.view.followMode`.
* jump to next/previous bookmark with commands `inlineBookmarks.jumpToNext`, `inlineBookmarks.jumpToPrevious` (assign your own keyboard shortcut in vscode preferences) 
* apply custom filter to bookmarks view with command `inlineBookmarks.setTreeViewFilterWords`.
  * empty list unsets filter
  * takes space- or semicolon-delimited list of regular expressions applied on the items label.
  * affects `jumpToNext`, `jumpToPrevious`
  
Buttons (left to right):
- Jump to previous bookmark.
- Jump to next bookmark.
- Filter bookmark view: the prompt accepts regular expressions. keep empty to disable filtering.
- Toggle: show bookmark for visible editors only.
- Quick Refresh: refreshes the bookmark from the internal cache.
- Scan Workspace for Bookmarks: scans all documents in the workspace for bookmark tags

 <img width="350" alt="inline_bookmarks_icon" src="https://user-images.githubusercontent.com/2865694/103533925-44b57f80-4e8e-11eb-8602-8fda358c8961.png">

Optionally, hide items from the bookmarks view that are excluded by a downstream `.gitignore` file (`inline-bookmarks.view.exclude.gitIgnore`; default: `false`; requires reload).

<img width="541" alt="image" src="https://user-images.githubusercontent.com/2865694/194845897-2146a70f-ac11-4ba9-8406-a1b582e9bf3c.png">

## FAQ

#### Q: Where do I find more settings?

Go to `code → preferences → Extensions: Inline Bookmarks`.

#### Q: How can I reset the extensions bookmark cache in case of permanent errors?

Bookmarks are cached in the vscode workspace. In case of permanent "ghost entries" or other errors you might want to try to execute the command: `inlineBookmarks.debug.state.reset`. This is going to reset the cache and allow the extension to populate it from scratch. Bookmarks are typically added as you go when opening new files in the editor. You can also make the extension scan the workspace for files containing Bookmarks. We don't do this automatically as it is quite resource intensive.

#### Q: How can I control which paths/file-extensions are being processed by the extension?

By default all paths are included (`inline-bookmarks.search.includes`) except the ones defined with `inline-bookmarks.search.excludes` (supports wildcard path globs).

Additionally, file-extensions configured with `inline-bookmarks.exceptions.file.extensions.ignore` will be excluded as well (prefer this over `search.excludes`). 

#### Q: How do I temporarily exempt certain trigger-words from being decorated?

See `inline-bookmarks.exceptions.words.ignore` (matches the beginning of the word).

#### Q: How can I define custom bookmark trigger-words/labels and colors?

The extension will search for all the default trigger-words configured with the extension. Note that these default trigger-words can be overriden (or removed). In addition, we give you complete freedom over any custom trigger-words you would want to configure. See example.

**Note**

* Existing words and styles can be overriden.
* `gutterIconColor` may be used to specify a custom icon color using any RGBA format. gutterIconColor will override gutterIconPath. See example below.
* (Deprecated) `gutterIconPath` may refer to only the four images provided with the extension right now: `images/bookmark-{red,green,blue,purple}.svg`. See example below. 
* You can assign multiple regex trigger words to a decoration style. See example.

**Example word mapping:** (accepts regular expressions; `\` needs to be encoded as `\\`)

```json
"inline-bookmarks.expert.custom.words.mapping": {
    "blue": ["@audit\\-info[ \\t\\n]"],
    "purple": ["@audit\\-issue[ \t\\n]"],
    "green": ["@audit\\-ok[ \\t\\n]"],
    "red": ["@audit[ \\t\\n]"],
    "warn": ["@warn[ \\t\\n]"] 
}
```

**Example style definition:** (all [vscode style properties](https://code.visualstudio.com/api/references/vscode-api#DecorationRenderOptions) are allowed)

```json
"inline-bookmarks.expert.custom.styles": {
    "default": {
        "gutterIconColor": "#157EFB",
        "overviewRulerColor": "rgba(21, 126, 251, 0.7)",
        "light": {
            "fontWeight": "bold"
        },
        "dark": {
            "color": "Chocolate"
        }
    },
    "red": {
        "gutterIconColor": "#F44336",
        "light": {
            "fontWeight": "bold"
        },
        "dark": {
            "color": "Chocolate"
        }
    },
    "blue": {
        "gutterIconColor": "#157EFB",
        "light": {
            "fontWeight": "bold"
        },
        "dark": {
            "color": "Chocolate"
        }
    },
    "green": {
        "gutterIconColor": "#2FCE7C",
        "light": {
            "fontWeight": "bold"
        },
        "dark": {
            "color": "Chocolate"
        }
    },
    "purple": {
        "gutterIconColor": "#C679E0",
        "light": {
            "fontWeight": "bold"
        },
        "dark": {
            "color": "Chocolate"
        }
    },
    "warn": {   // example custom style with yellow color
        "gutterIconColor": "#F4F400",
        "overviewRulerColor": "#F4F400B0",
        "light": {
            "fontWeight": "bold"
        },
        "dark": {
            "color": "Chocolate"
        }
    }
}
```

#### Q: How can I bind keys to `jumpToNext` and `jumpToPrevs` bookmark?

See https://code.visualstudio.com/docs/getstarted/keybindings.

#### Q: How can I change the bookmark-view's follow mode?

`inline-bookmarks.view.followMode` allows you to specify if you want to select the `nearest` bookmark relative to the current editor selection (default) or the next one (`chapter` mode).

#### Q: How can I make the bookmark-view's `jumpToNext` and `jumpToPrevs` behavior use the current selected line instead of the last selected bookmark?

`inline-bookmarks.view.lineMode` can be configured to `current-line`. Also see #40.

#### Q: How can I scan the workspace for bookmarks?


CHANGELOG.md
# Change Log

## 0.1.0
- fix: VSCode 1.72.0 introduced an incompatibility raising the error: 'Extension tintinweb.vscode-inline-bookmarks has provided an invalid tree item.' - #56 #57
- new: Optionally, hide items from the bookmarks view that are excluded by a downstream `.gitignore` file (requires reload)
  - option: `inline-bookmarks.view.exclude.gitIgnore` default: `false`

<img width="541" alt="image" src="https://user-images.githubusercontent.com/2865694/194845897-2146a70f-ac11-4ba9-8406-a1b582e9bf3c.png">



## 0.0.27
- fix: Command 'Select Bookmark' results in error 'filter is not a function' - #50 #51
- update: License 👉 [GPLv3](LICENSE)

## 0.0.26
- new: config setting: `inline-bookmarks.view.showVisibleFilesOnlyMode =  "allVisibleEditors" || "onlyActiveEditor"` to control the Inline Bookmarks View's "show visible editors only" behavior - #49
  - `allVisibleEditors` - default behavior. with "show visible editors only" and `allVisibleEditors` mode bookmarks for all visible editors are shown in the Inline Bookmarks View
  - `onlyActiveEditor` - with "show visible editors only" and `allVisibleEditors` mode only bookmarks for the currently selected/active editor are shown in the Inline Bookmarks View

- fix: default bookmark/tag regex to include all whitespaces (including CRLF). - #48


## 0.0.25
- fix: list bookmarks not working correctly - #43

## 0.0.24
- new: support for custom bookmarks colors. More information see [Readme.md/FAQ/Custom Styles and Mapping](./README.md#FAQ) Settings - see #30, #42 (thanks @rkodey)
- new: `inline-bookmarks.view.lineMode` allows to specify the behavior when jumping to the next/previous bookmark. Either use the current `selected-bookmark` or the `current-line` selected in the editor as the reference for the next jump. - see #40, #41 (thanks @rkodey)
- fixed: jump errors due to vscode api misuse - see #38, #39 (thanks @rkodey)

## 0.0.23
- new: Color indicators for the overview ruler `overviewRulerColor` for all tags - #35

## 0.0.22
- new: command `List Visible Bookmarks` prints bookmarks for currently visible editors (use `List Bookmarks` if you want to print all bookmarks)
- new: command `Select Visible Bookmark` that allows to quickly jump to bookmarks in currently visible editors (use `Select Bookmark` if you want to select from an unfiltered list)

## 0.0.21
- fixed: an update to vscode caused the list-view selection tracking to fail with "cannot read property '_uri'  of undefined" - see #33 

## 0.0.20
- new: new quick actions to bookmarks view - #20
  - Jump to Previous/Next, Filter View, Scan Workspace for Bookmarks (slow)
  - renamed "Set View Filter" to "Filter View ..."
  - updated "Filter View ..." prompt with a note that an empty filter disables filtering
  
Buttons (left to right):
- Jump to previous bookmark.
- Jump to next bookmark.
- Filter bookmark view: the prompt accepts regular expressions. keep empty to disable filtering.
- Toggle: show bookmark for visible editors only.
- Quick Refresh: refreshes the bookmark from the internal cache.
- Scan Workspace for Bookmarks: scans all documents in the workspace for bookmark tags

 ![image](https://user-images.githubusercontent.com/2865694/103533925-44b57f80-4e8e-11eb-8602-8fda358c8961.png)

 

## 0.0.19
- new: command to list all bookmarks (text)
- new: experimental command to scan workspace for bookmarks
  - note: don't use this on large workspaces. if it breaks use the `onCommand:inlineBookmarks.debug.state.reset` command from the command palette
- fixed: listview filter not working correctly - #24

## 0.0.18
- fixed: extension not fully registered if vscode is slow or editor not set
  - may manifest as cross-extension incompatibility as reported with #17

## 0.0.17
- new: added functionality to quickly filter the list of bookmarks `inlineBookmarks.showSelectBookmark` #11
- fixed: multiple issues where tree items or bookmarks where undefined #16
- new: debug command to reset the stored bookmarks read from workspace. in case the workspace is corrupt. `inlineBookmarks.debug.state.reset` #16

## 0.0.16
- new: added commands to jump to next/previous bookmark: `inlineBookmarks.jumpToNext` `inlineBookmarks.jumpToPrevious` #12
  - **Note**: keyboard shortcuts can be configured in `code -> preferences -> keyboard shortcuts` #15
- new: added functionality to filter the bookmark view #14
- fixed: remove `jumpToLine` from command menu #13

## 0.0.15
- fixed: treeview item id(0) error
- new: set `inline-bookmarks.view.words.hide` to hide tags/trigger words in bookmarks view (only for entries with additional text)

## 0.0.14
- fixed: wrong treeview item selected #7

## 0.0.13
- allow to specify bookmarks view follow behavior: `inline-bookmarks.view.followMode`
  - `nearest` - highlights the nearest bookmark (relative distance)
  - `chapter` - highlights the most recent bookmark or the bookmark that is on the current line.

## 0.0.12

- new: sync the bookmarks view with the editor click action (highlight nearest)
- fixed: sort bookmarks view by line number


src/features/inlineBookmarks.js
'use strict';
/** 
 * @author github.com/tintinweb
 *  
 * 
 * 
 * */
/** imports */
const vscode = require('vscode');
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const settings = require('../settings');
const os = require("os");

class Commands {
    constructor(controller) {
        this.controller = controller;
    }

    refresh() {
        Object.keys(this.controller.bookmarks).forEach(uri => {
            vscode.workspace.openTextDocument(vscode.Uri.parse(uri)).then(document => {
                this.controller.updateBookmarks(document);
            });
        }, this);
    }

    showSelectBookmark(filter, placeHolder) {

        let entries = [];
        Object.keys(this.controller.bookmarks).forEach(uri => {
            let resource = vscode.Uri.parse(uri).fsPath;
            let fname = path.parse(resource).base;

            if(filter && !filter(resource)){
                return;
            }

            Object.keys(this.controller.bookmarks[uri]).forEach(cat => {
                this.controller.bookmarks[uri][cat].forEach(b => {
                    entries.push({
                        label: b.text,
                        description: fname,
                        target: new vscode.Location(resource, b.range)
                    });
                });
            });

        }, this);

        vscode.window.showQuickPick(entries, { placeHolder: placeHolder || 'Select bookmarks' }).then(item => {
            vscode.commands.executeCommand("inlineBookmarks.jumpToRange", item.target.uri, item.target.range);
        });
    }

    showSelectVisibleBookmark() {
        let visibleEditorUris = vscode.window.visibleTextEditors.map(te => te.document.uri.fsPath);
        this.showSelectBookmark((resFsPath) => visibleEditorUris.includes(resFsPath), "Select visible bookmarks");
    }

    showListBookmarks(filter) {

        if (!vscode.window.outputChannel) {
            vscode.window.outputChannel = vscode.window.createOutputChannel('inlineBookmarks');
        }

        if (!vscode.window.outputChannel) return;
        vscode.window.outputChannel.clear();

        let entries = [];
        Object.keys(this.controller.bookmarks).forEach(uri => {
            let resource = vscode.Uri.parse(uri).fsPath;
            let fname = path.parse(resource).base;

            if(filter && !filter(resource)){
                return;
            }

            Object.keys(this.controller.bookmarks[uri]).forEach(cat => {
                this.controller.bookmarks[uri][cat].forEach(b => {
                    entries.push({
                        label: b.text,
                        description: fname,
                        target: new vscode.Location(resource, b.range)
                    });
                });
            });

        }, this);

        if (entries.length === 0) {
            vscode.window.showInformationMessage('No results');
            return;
        }

        entries.forEach(function (v, i, a) {
            var patternA = '#' + (i + 1) + '\t' + v.target.uri + '#' + (v.target.range.start.line + 1);
            var patternB = '#' + (i + 1) + '\t' + v.target.uri + ':' + (v.target.range.start.line + 1) + ':' + (v.target.range.start.character + 1);
            var patterns = [patternA, patternB];

            var patternType = 0;
            if (os.platform() == "linux") {
                patternType = 1;
            }
            patternType = +!patternType;

            vscode.window.outputChannel.appendLine(patterns[patternType]);
            vscode.window.outputChannel.appendLine('\t' + v.label + '\n');
        });
        vscode.window.outputChannel.show();
    }

    showListVisibleBookmarks() {
        let visibleEditorUris = vscode.window.visibleTextEditors.map(te => te.document.uri.fsPath);
        this.showListBookmarks((resFsPath) => visibleEditorUris.includes(resFsPath));
    }

    scanWorkspaceBookmarks() {

        function arrayToSearchGlobPattern(config) {
            return Array.isArray(config) ?
                '{' + config.join(',') + '}'
                : (typeof config == 'string' ? config : '');
        }

        var includePattern = arrayToSearchGlobPattern(settings.extensionConfig().search.includes) || '{**/*}';
        var excludePattern = arrayToSearchGlobPattern(settings.extensionConfig().search.excludes);
        var limit = settings.extensionConfig().search.maxFiles;

        let that = this;
    
        vscode.workspace.findFiles(includePattern, excludePattern, limit).then(function (files) {
    
            if (!files || files.length === 0) {
                console.log('No files found' );
                return;
            }
    
            var totalFiles = files.length;

            for (var i = 0; i < totalFiles; i++) {
    
                vscode.workspace.openTextDocument(files[i]).then((document) => {
                    that.controller.updateBookmarks(document);
                    //NOP
                }, (err) => {
                    console.error(err);
                });
    
            }
            
        }, (err) => {
            console.error(err);
        });
    }

}

class InlineBookmarksCtrl {

    constructor(context) {
        this.context = context;
        this.styles = this._reLoadDecorations();
        this.words = this._reLoadWords();

        this.commands = new Commands(this);

        this.bookmarks = {};  // {file: {bookmark}}
        this.loadFromWorkspace();
    }

    /** -- public -- */

    hasBookmarks() {
        return !!this.bookmarks;
    }

    async decorate(editor) {
        if (!editor || !editor.document /*|| editor.document.fileName.startsWith("extension-output-")*/) return; //decorate list of inline comments

        this._clearBookmarksOfFile(editor.document);

        if (this._extensionIsBlacklisted(editor.document.fileName)) return;

        for (var style in this.words) {
            if (!this.words.hasOwnProperty(style) || this.words[style].length == 0 || this._wordIsOnIgnoreList(this.words[style])) {
                continue;
            }
            this._decorateWords(editor, this.words[style], style, editor.document.fileName.startsWith("extension-output-")); //dont add to bookmarks if we're decorating an extension-output
        }

        this.saveToWorkspace(); //update workspace
    }

    async updateBookmarks(document) {
        if (!document || document.fileName.startsWith("extension-output-")) return;

        this._clearBookmarksOfFile(document);

        if (this._extensionIsBlacklisted(document.fileName)) return;

        for (var style in this.words) {
            if (!this.words.hasOwnProperty(style) || this.words[style].length == 0 || this._wordIsOnIgnoreList(this.words[style])) {
                continue;
            }
            this._updateBookmarksForWordAndStyle(document, this.words[style], style);
        }

        this.saveToWorkspace(); //update workspace
    }

    /** -- private -- */

    _extensionIsBlacklisted(fileName) {
        let ignoreList = settings.extensionConfig().exceptions.file.extensions.ignore;
        if (!ignoreList || ignoreList.length === 0) return false;
        return this._commaSeparatedStringToUniqueList(ignoreList).some(ext => fileName.endsWith(ext.trim()));
    }

    _wordIsOnIgnoreList(word) {
        let ignoreList = settings.extensionConfig().exceptions.words.ignore;
        return this._commaSeparatedStringToUniqueList(ignoreList).some(ignoreWord => word.startsWith(ignoreWord.trim()));
    }

    _commaSeparatedStringToUniqueList(s) {
        if (!s) return [];
        return [...new Set(s.trim().split(',').map(e => e.trim()).filter(e => e.length))];
    }

    async _decorateWords(editor, words, style, noAdd) {
        const decoStyle = this.styles[style].type || this.styles['default'].type;

        let locations = this._findWords(editor.document, words);
        editor.setDecorations(decoStyle, locations);  // set decorations

        if (locations.length && !noAdd)
            this._addBookmark(editor.document, style, locations);
    }

    async _updateBookmarksForWordAndStyle(document, words, style) {

        let locations = this._findWords(document, words);

        if (locations.length)
            this._addBookmark(document, style, locations);
    }

    _findWords(document, words) {
        const text = document.getText();
        var locations = [];

        words.forEach(function (word) {

            var regEx = new RegExp(word, "g");
            let match;
            while (match = regEx.exec(text)) {

                var startPos = document.positionAt(match.index);
                var endPos = document.positionAt(match.index + match[0].trim().length);

                var fullLine = document.getWordRangeAtPosition(startPos, /(.+)$/);

                var decoration = {
                    range: new vscode.Range(startPos, endPos),
                    text: document.getText(new vscode.Range(startPos, fullLine.end))
                };

                locations.push(decoration);
            }
        });

        return locations;
    }

    _clearBookmarksOfFile(document) {
        let filename = document.uri;
        if (!this.bookmarks.hasOwnProperty(filename)) return;
        delete this.bookmarks[filename];
    }

    _clearBookmarksOfFileAndStyle(document, style) {
        let filename = document.uri;
        if (!this.bookmarks.hasOwnProperty(filename)) return;
        delete this.bookmarks[filename][style];
    }

    _addBookmark(document, style, locations) {
        let filename = document.uri;
        if (!this.bookmarks.hasOwnProperty(filename)) {
            this.bookmarks[filename] = {};
        }
        this.bookmarks[filename][style] = locations;
    }

    _reLoadWords() {
        let defaultWords = {  // style: arr(regexWords)
            "blue": this._commaSeparatedStringToUniqueList(settings.extensionConfig().default.words.blue),
            "purple": this._commaSeparatedStringToUniqueList(settings.extensionConfig().default.words.purple),
            "green": this._commaSeparatedStringToUniqueList(settings.extensionConfig().default.words.green),
            "red": this._commaSeparatedStringToUniqueList(settings.extensionConfig().default.words.red)
        };

        return { ...defaultWords, ...settings.extensionConfig().expert.custom.words.mapping };
    }

    _getBookmarkDataUri(color) {
        return vscode.Uri.parse(
            "data:image/svg+xml," +
            encodeURIComponent(`<svg version="1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 48 48" enable-background="new 0 0 48 48"><path fill="${color}" d="M37,43l-13-6l-13,6V9c0-2.2,1.8-4,4-4h18c2.2,0,4,1.8,4,4V43z"/></svg>`)
        );
    }

    _getDecorationStyle(decoOptions) {
        return { type: vscode.window.createTextEditorDecorationType(decoOptions), options: decoOptions };
    }

    _getDecorationDefaultStyle(color) {
        return this._getDecorationStyle({
            "gutterIconPath": this._getBookmarkDataUri(color),
            "overviewRulerColor": color+"B0",   // this is safe/suitable for the defaults only.  Custom ruler color is handled below.
            "light": {
                "fontWeight": "bold"
            },
            "dark": {
                "color": "Chocolate"
            }
        })
    }

    _reLoadDecorations() {
        const blue      = '#157EFB';
        const green     = '#2FCE7C';
        const purple    = '#C679E0';
        const red       = '#F44336';
        let styles      = {
            "default":  this._getDecorationDefaultStyle(blue),
            "red":      this._getDecorationDefaultStyle(red),
            "blue":     this._getDecorationDefaultStyle(blue),
            "green":    this._getDecorationDefaultStyle(green),
            "purple":   this._getDecorationDefaultStyle(purple)
        };

        let customStyles = settings.extensionConfig().expert.custom.styles;

        for (var decoId in customStyles) {

            if (!customStyles.hasOwnProperty(decoId)) {
                continue;
            }

            let decoOptions = { ...customStyles[decoId] };

            // default to blue if neither an icon path nor an icon color is specified
            if (!decoOptions.gutterIconPath) {
                decoOptions.gutterIconColor = decoOptions.gutterIconColor || blue;
            }

            //apply icon color if provided, otherwise fix the path
            decoOptions.gutterIconPath = decoOptions.gutterIconColor ? this._getBookmarkDataUri(decoOptions.gutterIconColor) : this.context.asAbsolutePath(decoOptions.gutterIconPath);

            //overview
            if (decoOptions.overviewRulerColor) {
                decoOptions.overviewRulerLane = vscode.OverviewRulerLane.Full;
            }
            //background color
            if (decoOptions.backgroundColor) {
                decoOptions.isWholeLine = true;
            }
            styles[decoId] = this._getDecorationStyle(decoOptions);
        }

        return styles;
    }

    _isWorkspaceAvailable() {
        //single or multi root
        return vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length >= 1;
    }

    resetWorkspace() {
        if (!this._isWorkspaceAvailable()) return; //cannot save
        this.context.workspaceState.update("bookmarks.object", "{}");
    }

    saveToWorkspace() {
        if (!this._isWorkspaceAvailable()) return; //cannot save
        this.context.workspaceState.update("bookmarks.object", JSON.stringify(this.bookmarks));
    }

    loadFromWorkspace() {
        if (!this._isWorkspaceAvailable()) return; //cannot load
        this.bookmarks = JSON.parse(this.context.workspaceState.get("bookmarks.object", "{}"));

        //remove all non existing files
        Object.keys(this.bookmarks).forEach(filepath => {
            if (!fs.existsSync(vscode.Uri.parse(filepath).fsPath)) {
                delete this.bookmarks[filepath];
                return;
            }

            Object.keys(this.bookmarks[filepath]).forEach(cat => {
                //for each category
                this.bookmarks[filepath][cat] = this.bookmarks[filepath][cat].map(decoObject => {
                    //fix - rebuild range object (it is expected by other functions)
                    decoObject.range = new vscode.Range(decoObject.range[0].line, decoObject.range[0].character, decoObject.range[1].line, decoObject.range[1].character);
                    return decoObject;
                });
            });
        });
    }
}

const NodeType = {
    FILE: 1,
    LOCATION: 2
};


class InlineBookmarksDataModel {

    /** treedata model */

    constructor(controller) {
        this.controller = controller;
    }

    getRoot() {  /** returns element */
        let fileBookmarks = Object.keys(this.controller.bookmarks);
        
        if (settings.extensionConfig().view.showVisibleFilesOnly) {

            let visibleEditorUris;
            if(settings.extensionConfig().view.showVisibleFilesOnlyMode === "onlyActiveEditor") {
                visibleEditorUris = [vscode.window.activeTextEditor.document.uri.path];
            } else {
                visibleEditorUris = vscode.window.visibleTextEditors.map(te => te.document.uri.path);
            }

            fileBookmarks = fileBookmarks.filter(v => visibleEditorUris.includes(vscode.Uri.parse(v).path));
        }

        return fileBookmarks.sort().map(v => {
            return {
                resource: vscode.Uri.parse(v),
                tooltip: v,
                name: v,
                type: NodeType.FILE,
                parent: null,
                iconPath: vscode.ThemeIcon.File,
                location: null
            };
        });
    }

    getChildren(element) {
        switch (element.type) {
            case NodeType.FILE:
                let bookmarks = Object.keys(this.controller.bookmarks[element.name]).map(cat => {
                    //all categories
                    return this.controller.bookmarks[element.name][cat].map(v => {
                        let location = new vscode.Location(element.resource, v.range);
                        return {
                            resource: element.resource,
                            location: location,
                            label: v.text.trim(),
                            name: v.text.trim(),
                            type: NodeType.LOCATION,
                            category: cat,
                            parent: element,
                            iconPath: this.controller.styles[cat].options.gutterIconPath
                        };
                    });
                }).flat(1);

                return bookmarks.sort((a, b) => a.location.range.start.line - b.location.range.start.line);
                break;
        }
    }

    /**
    Find previous and next of element (for goto_next, goto_previous)

    requires current element from tree
    */
    getNeighbors(element) {
        let ret = { previous: null, next: null };
        let parent = element.parent;
        if (!parent) {
            //fake the parent
            parent = { ...element };  //use parent or derive it from bookmark
            parent.type = NodeType.FILE;
            parent.name = element.resource;
        }

        //get all children
        let bookmarks = this.getChildren(parent);

        //lets track if we're at our element.
        let gotElement = false;

        for (let b of bookmarks) {
            // find element in list, note prevs, next
            if (!gotElement && JSON.stringify(b.location) == JSON.stringify(element.location)) {
                gotElement = true;
                continue;
            }
            if (!gotElement) {
                ret.previous = b;
            } else {
                ret.next = b;
                break;
            }
        }

        return ret;
    }

}

class InlineBookmarkTreeDataProvider {

    constructor(inlineBookmarksController) {
        this._onDidChangeTreeData = new vscode.EventEmitter();
        this.onDidChangeTreeData = this._onDidChangeTreeData.event;

        this.controller = inlineBookmarksController;
        this.model = new InlineBookmarksDataModel(inlineBookmarksController);

        this.filterTreeViewWords = [];
        this.gitIgnoreHandler = undefined;
    }

    /** events */

    /** methods */

    getChildren(element) {  
        return this._filterTreeView(element ? this.model.getChildren(element) : this.model.getRoot());
    }

    getParent(element) {
        return element ? element.parent : element;
    }

    getTreeItem(element) {
        if (!element) {
            return element; // undef
        }
        let item = new vscode.TreeItem(
            this._formatLabel(element.label), 
            element.type == NodeType.LOCATION ? 0 : settings.extensionConfig().view.expanded ? vscode.TreeItemCollapsibleState.Expanded : vscode.TreeItemCollapsibleState.Collapsed
            );
        item.id = element.type == NodeType.LOCATION ? this._getId(element.location) : this._getId(element.resource);
        item.resourceUri = element.resource;
        item.iconPath = element.iconPath;
        item.command = element.type == NodeType.LOCATION && element.location ? {
            command: 'inlineBookmarks.jumpToRange',
            arguments: [element.location.uri, element.location.range],
            title: 'JumpTo'
        } : 0;
        return item;
    }

    /* 
    Hash object to unique ID.
    */
    _getId(o) {
        return crypto.createHash('sha1').update(JSON.stringify(o)).digest('hex');
    }

    _formatLabel(label) {
        if (!settings.extensionConfig().view.words.hide || !label) {
            return label;
        }
        let words = Object.values(this.controller.words).flat(1);
        return words.reduce((prevs, word) => prevs.replace(new RegExp(word, "g"), ""), label);  //replace tags in matches.
    }

    _filterTreeView(elements) {

        if(this.gitIgnoreHandler && this.gitIgnoreHandler.filter){
            elements = elements.filter(e => this.gitIgnoreHandler.filter(e.resource));
        }

        if (this.filterTreeViewWords && this.filterTreeViewWords.length) {
            elements = elements.filter(e => this.filterTreeViewWords.some(rx => new RegExp(rx, 'g').test(e.label)));
        }

        return elements;
    }
    /** other methods */

    setTreeViewFilterWords(words) {
        this.filterTreeViewWords = words;
    }

    setTreeViewGitIgnoreHandler(gi) {
        this.gitIgnoreHandler = gi;
    }

    refresh() {
        this._onDidChangeTreeData.fire();
    }
}


module.exports = {
    InlineBookmarksCtrl: InlineBookmarksCtrl,
    InlineBookmarkTreeDataProvider: InlineBookmarkTreeDataProvider,
    NodeType: NodeType
};
